docs/config.rst
Configuration
================================

coco's main configuration file has the following options. It can be passed to coco with `coco[d] -c path/to/coco.conf`.

log_level : `str`
    The global log level. Can be one of `CRITICAL`, `ERROR`, `WARNING`, `INFO` or `DEBUG`. Default `INFO`.
host: `str`
    Host the coco server (`cocod`) is run on.
port: `int`
    Port the coco server (`cocod`) should be run on. Default `12055`.
metrics_port: `int`
    Port cocod should run its metrics server on. Default `9090`.
endpoint_dir: `str`
    Path where endpoint config files are located.
n_workers: `int`
    Number of sanic workers to start for the frontend of `cocod`. Default `1`.
session_limit: `int`
    Maximum number of tasks being executed concurrently by request forwarder. A higher number will use more memory. Default `1000`.
blocklist_path: `str`
    Path to persistent blocklist storage file. Default `/var/lib/coco/blocklist.json`.
storage_path: `str`
    Path to persistent state storage. Default `/var/lib/coco/state/`.
groups:
    Groups of nodes that are managed by coco. Contains key-value pairs where the key is
    the name of the group (`str`) and the value is a list of `str` in the format
    `host:port`.

    Example:

.. code-block:: yaml

    groups:
        gps_server:
            - carillon.chime:54321
        cluster:
            - localhost:12050
            - localhost:12000
        receiver_nodes:
            - recv1:12048
            - recv2:12048
        all:
            - localhost:12050
            - localhost:12000
            - recv1:12048
            - recv2:12048

load_state:
    Initial cluster state.
    Contains key-value pairs that are state paths and file system paths. The files
    located at the latter locations are loaded into the prior parts of the state when
    resetting the state. Default: `None`.

    Example:

.. code-block:: yaml

    load_state:
        cluster: "../conf/gpu.yaml"
        receiver: "../conf/recv.yaml"

slack_token: `str`
    Slack authorization token. Default: `None`.
slack_rules: list
    Rules for dispatching logging messages to slack.
    These specify the logger path, the minimum level it applies to and the
    slack channel the messages should go to. Default: `None`.

    Example:

.. code-block:: yaml

    slack_rules:

        - logger: coco
          level: WARNING
          channel: coco-alerts

        - logger: coco.endpoint.update-tracking-pointing-0
          level: INFO
          channel: pulsar-timing-ops

queue_length: int
    Length of the endpoint request queue between front- end backend, managed by redis.
    Default `0`.
timeout: str
    Time before requests sent to nodes time out.
    A string representing a timedelta in the form `<int>h`, `<int>m`,
    `<int>s` or a combination of the three. Default `10s`.
frontend_timeout: str
    Time before requests sent to coco time out.
    This value should depend on how many layers your configuration files have. If a call
    to a coco endpoint could take longer than this value, because it triggers many
    layered forward calls you should increase this.
    A string representing a timedelta in the form `<int>h`, `<int>m`,
    `<int>s` or a combination of the three. Default `10m`.
exclude_from_reset: list
    A list of strings that are state paths to be excluded from state resets. Default:
    `None`


CHANGELOG.md
### [2020.10.0](https://github.com/chime-experiment/coco/compare/2020.06.0...2020.10.0) (2020-09-30)


### Bug Fixes

* **endpoint:** Allow 'set_state' value to have initial None type ([29442fa](https://github.com/chime-experiment/coco/commit/29442fa629716d8a665d9f04e2c1e905af4f16d1))
* **StateReplyCheck:** don't initialize as dict ([8297697](https://github.com/chime-experiment/coco/commit/8297697649e2eeec9ce8072db9713b49ef438f3b))
* use 127.0.0.1 instead of localhost ([44b9069](https://github.com/chime-experiment/coco/commit/44b9069e20eb010db26160afeff78fdc1efed15d)), closes [#213](https://github.com/chime-experiment/coco/issues/213)


### Features

* **StateReplyCheck:** speed up state diffs ([01914d2](https://github.com/chime-experiment/coco/commit/01914d2a9d708fbdd971d5722c84bf6747b30689)), closes [#161](https://github.com/chime-experiment/coco/issues/161)


### Styles

* rename blacklist -> blocklist ([8098e28](https://github.com/chime-experiment/coco/commit/8098e2859b5696315ccc0b4811d24ff02b64cb41))


### BREAKING CHANGES

* The command blacklist for the coco command line client is now called
blocklist.



### [2020.06.0](https://github.com/chime-experiment/coco/compare/2020.04.0...2020.06.0) (2020-06-11)


### Bug Fixes

* **client:** typo in update-blacklist help text ([a9531f6](https://github.com/chime-experiment/coco/commit/a9531f6dfe7878554c34df45cbded65477b9244b))
* **endpoint:** pass host to metric.get(queue_length) ([53ed20c](https://github.com/chime-experiment/coco/commit/53ed20cb7d097712645905f67e596642ca9b6396)), closes [#194](https://github.com/chime-experiment/coco/issues/194)
* **worker:** ndle redis server timing out ([011678b](https://github.com/chime-experiment/coco/commit/011678be1d299047d16b2521fd3b122c7293f44d)), closes [#203](https://github.com/chime-experiment/coco/issues/203)


### Features

* **cocod:** add option --check-config ([f859728](https://github.com/chime-experiment/coco/commit/f859728e1ff155c62be5c10a40626a86023133bf))
* **master:** add defaults to Master.__init__ arguments ([facb258](https://github.com/chime-experiment/coco/commit/facb258e6dbf16852ab68cec55a593d86ddcdee1))
* **state:** commands to save and load states ([97a0521](https://github.com/chime-experiment/coco/commit/97a05218cefc83c59c256286537773dd1cbbd0eb))


### BREAKING CHANGES

* **state:** The config variable 'storage_path' is not a file anymore, but a
directory. The state is saved there under the name 'active' and saved
states are added here as well. If you are updating your coco instance,
you have to manually move the persistent state from 'storage_path' to
'storage_path'/active.



# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2020.04.0](https://github.com/chime-experiment/coco/compare/2019.12.0...2020.04.0) (2020-04-10)


### Features

* **coco client:** Improve output & show progress ([0315bae](https://github.com/chime-experiment/coco/commit/0315bae)), closes [#186](https://github.com/chime-experiment/coco/issues/186)
* **cocod:** add option --reset ([9a7f9a7](https://github.com/chime-experiment/coco/commit/9a7f9a7)), closes [#184](https://github.com/chime-experiment/coco/issues/184)

## 2019.12.0 (2019-12-09)


### Bug Fixes


* **state:** ignore excluded states if they don't exist ([bc8d8a0](https://github.com/chime-experiment/coco/commit/bc8d8a0))
* **state:** when hashing recurse on list elements ([cbec4df](https://github.com/chime-experiment/coco/commit/cbec4df))

## 2019.11.0 (2019-11-21)


### Bug Fixes

* **master:** disable sanic debug mode ([17c0c08](https://github.com/chime-experiment/coco/commit/17c0c08))


### Features

* **metrics** add metric for tracking how long a request is waiting in the queue ([e2fb5cf](https://github.com/chime-experiment/coco/commit/e2fb5cf))
* **state:** add option: exclude_from_reset ([d9792ec](https://github.com/chime-experiment/coco/commit/d9793ec))


## 2019.10.0 (2019-10-22)


### Bug Fixes

* **request_forwarder:** copy request for coco forwards ([c04fe33](https://github.com/chime-experiment/coco/commit/c04fe33)), closes [#171](https://github.com/chime-experiment/coco/issues/171)
* **slack:** set slack message timeout ([2636da7](https://github.com/chime-experiment/coco/commit/2636da7)), closes [#165](https://github.com/chime-experiment/coco/issues/165)
* **state:** handle errors in _find(path) ([5bcb14c](https://github.com/chime-experiment/coco/commit/5bcb14c))
* **state:** get hash of msgpacked state ([c859d1c](https://github.com/chime-experiment/coco/commit/c859d1c)), closes [kotekan/kotekan#477](https://github.com/kotekan/kotekan/issues/477)
* **worker:** answer with 404 if endpoint unknown ([aaae74a](https://github.com/chime-experiment/coco/commit/aaae74a))


### Features

* **slack:** print error if slack logging failed ([94009a2](https://github.com/chime-experiment/coco/commit/94009a2))
* **state:** make state persistent ([ea36e5a](https://github.com/chime-experiment/coco/commit/ea36e5a)), closes [#85](https://github.com/chime-experiment/coco/issues/85)


coco/config.py
"""For configuring coco from the config files.

Configuration file search order:

- `/etc/coco/coco.conf`
- `/etc/xdg/coco/coco.conf`
- `~/.config/coco/coco.conf`
- `COCO_CONFIG_FILE` environment variable

This is in order of increasing precendence, with options in later files
overriding those in earlier entries. Configuration is merged recursively by
`merge_dict_tree`.

Example config:

.. codeblock:: yaml

    # Configure the standard logging level
    log_level: "DEBUG"

    # Directory to load endpoint configuration from
    endpoint_dir: '../conf/endpoints'

    # Host and port to run the coco server
    host: localhost
    port: 12055

    # Port for prometheus metrics
    metrics_port: 12056

    # Number of workers that will process and forward requests
    n_workers: 2

    # Time before requests sent to nodes time out. Needs to be a string representing a timedelta in
    # the form `<int>h`, `<int>m`, `<int>s` or a combination of the three.
    timeout: 10s

    # Time before requests sent to coco time out.
    # This value should depend on how many layers your configuration files have. If a call to a
    # coco endpoint could take longer than this value, because it triggers many layered forward
    # calls you should increase this.
    # Needs to be a string representing a timedelta in
    # the form `<int>h`, `<int>m`, `<int>s` or a combination of the three.
    frontend_timeout: 10m

    # Groups of nodes that are managed by coco
    groups:
        gps_server:
            - carillon.chime:54321
        cluster:
            - localhost:12050
            - localhost:12000
        receiver_nodes:
            - recv1:12048
            - recv2:12048
        all:
            - localhost:12050
            - localhost:12000
            - recv1:12048
            - recv2:12048

    # Should we use (and where should we find) comet for tracking the coco config
    comet_broker:
        enabled: True
        host: recv1
        port: 12050

    # Initial cluster state
    load_state:
        cluster: "../conf/gpu.yaml"
        receiver: "../conf/recv.yaml"

    # Slack authorization token
    slack_token: "slack_bot_token"

    # Rules for dispatching logging messages to slack
    # These specify the logger path, the minimum level it applies to and the
    # slack channel the messages should go to.
    slack_rules:

        - logger: coco
          level: WARNING
          channel: coco-alerts

        - logger: coco.endpoint.update-pulsar-pointing-0
          level: INFO
          channel: pulsar-timing-ops

    # State paths to be excluded from reset.
    exclude_from_reset:
        - this/should/be/preserved
        - this_too
"""
import logging
import os
from pathlib import Path

import yaml


logger = logging.getLogger(__name__)

# TODO: pretty much all logging messages config out of this module are ignored
# as the default level has not yet been applied, some workaround should be
# figured out. For the moment, just uncomment the line below
# logging.getLogger().setLevel(logging.DEBUG)


class DefaultValue:
    """Tag a config node with a default value."""

    def __init__(self, value):
        self.value = value


class RequiredValue:
    """Tag a config node as being required."""

    pass


_config_skeleton = {
    "host": RequiredValue(),
    "port": DefaultValue(12055),
    "metrics_port": DefaultValue(9090),
    "log_level": DefaultValue("INFO"),
    "endpoint_dir": RequiredValue(),
    "n_workers": DefaultValue(1),
    "session_limit": DefaultValue(1000),
    "blocklist_path": DefaultValue("/var/lib/coco/blocklist.json"),
    "storage_path": DefaultValue("/var/lib/coco/state/"),
    "groups": RequiredValue(),
    "load_state": DefaultValue({}),
    "slack_token": DefaultValue(None),
    "slack_rules": DefaultValue([]),
    "queue_length": DefaultValue(0),
    "timeout": DefaultValue("10s"),
    "frontend_timeout": DefaultValue("10m"),
    "exclude_from_reset": DefaultValue([]),
}


def load_config(path=None):
    """Find and load the configuration from a file."""

    # Initialise with the default configuration
    config = _config_skeleton.copy()

    # Construct the configuration file path
    config_files = [
        "/etc/coco/coco.conf",
        "/etc/xdg/coco/coco.conf",
        "~/.config/coco/coco.conf",
    ]

    if "COCO_CONFIG_FILE" in os.environ:
        config_files.append(os.environ["COCO_CONFIG_FILE"])

    if path is not None:
        config_files.append(path)

    any_exist = False

    for cfile in config_files:

        # Expand the configuration file path
        absfile = Path(cfile).expanduser().resolve()

        if not absfile.exists():
            logger.debug(f"Could not find config file {absfile}")
            continue

        any_exist = True

        logger.info(f"Loading config file {cfile}")

        with absfile.open("r") as fh:
            conf = yaml.safe_load(fh)

        config = merge_dict_tree(config, conf)

    if not any_exist:
        raise RuntimeError("No configuration files available.")

    _validate_and_resolve(config)

    _load_endpoint_config(config)

    return config


def merge_dict_tree(a, b):
    """Merge two dictionaries recursively.

    The following rules applied:

      - Dictionaries at each level are merged, with `b` updating `a`.
      - Lists at the same level are combined, with that in `b` appended to `a`.
      - For all other cases, scalars, mixed types etc, `b` replaces `a`.

    Parameters
    ----------
    a, b : dict
        Two dictionaries to merge recursively. Where there are conflicts `b`
        takes preference over `a`.

    Returns
    -------
    c : dict
        Merged dictionary.
    """

    # Different types should return b
    if type(a) != type(b):
        return b

    # From this point on both have the same type, so we only need to check
    # either a or b.
    if isinstance(a, list):
        return a + b

    # Dict's should be merged recursively
    if isinstance(a, dict):
        keys_a = set(a.keys())
        keys_b = set(b.keys())

        c = {}

        # Add the keys only in a...
        for k in keys_a - keys_b:
            c[k] = a[k]

        # ... now the ones only in b
        for k in keys_b - keys_a:
            c[k] = b[k]

        # Recursively merge any common keys
        for k in keys_a & keys_b:
            c[k] = merge_dict_tree(a[k], b[k])

        return c

    # All other cases (scalars etc) we should favour b
    return b


def _validate_and_resolve(config):
    """Check that all required values are present and resolve default values."""

    stack = [("", config, "", None)]

    missing_values = []

    while stack:
        key, value, prefix, parent = stack.pop()

        # If node is a dict, add all child entries onto the stack for processing
        if isinstance(value, dict):
            for k, v in value.items():
                stack.append((k, v, f"{prefix}/{key}", value))

        # Replace default values
        if isinstance(value, DefaultValue):
            parent[key] = value.value

        if isinstance(value, RequiredValue):
            missing_values.append(f"{prefix}/{key}")

    if missing_values:
        for path in missing_values:
            logger.error(f'Resolved config missing required entry "{path}".')
        raise RuntimeError("Config missing required values.")


def _load_endpoint_config(config):
    """Load the endpoint config.

    The config is injected into the passed in config object.
    """

    config["endpoints"] = []

    endpoint_dir = Path(config["endpoint_dir"])

    for endpoint_file in endpoint_dir.iterdir():
        # Only accept files ending in .conf as endpoint configs.
        # Endpoint config files starting with an underscore (_) are disabled.
        if endpoint_file.suffix == ".conf" and not endpoint_file.name.startswith("_"):
            logger.debug(f"Loading endpoint config {endpoint_file}.")

            # Remove .conf from the config file name to get the name of the endpoint
            name = endpoint_file.stem

            with endpoint_file.open("r") as fh:
                try:
                    conf = yaml.safe_load(fh)
                except yaml.YAMLError as exc:
                    logger.error(f"Failure reading YAML file {endpoint_file}: {exc}")

            conf["name"] = name

            # TODO: validate the endpoint config in here
            config["endpoints"].append(conf)
